apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -600,53 +650,67 @@ deliver(ClientInfo, Delivers, Session0) ->
 
 -spec handle_timeout(clientinfo(), _Timeout, session()) ->
     {ok, replies(), session()} | {ok, replies(), timeout(), session()}.
-handle_timeout(ClientInfo, ?TIMER_PULL, Session0) ->
-    {Publishes, Session1} =
-        case ?IS_REPLAY_ONGOING(Session0) of
-            false ->
-                drain_buffer(fetch_new_messages(Session0, ClientInfo));
-            true ->
-                {[], Session0}
-        end,
-    Timeout =
-        case Publishes of
-            [] ->
-                get_config(ClientInfo, [idle_poll_interval]);
-            [_ | _] ->
-                0
-        end,
-    Session = emqx_session:ensure_timer(?TIMER_PULL, Timeout, Session1),
-    {ok, Publishes, Session};
+handle_timeout(_ClientInfo, ?TIMER_PULL, Session0) ->
+    %% Pull circuit loop:
+    ?tp(debug, sessds_pull, #{}),
+    Session1 = Session0#{?TIMER_PULL := undefined},
+    case drain_buffer(Session1) of
+        {[], Session} ->
+            {ok, [], push_now(Session)};
+        {Publishes, Session} ->
+            {ok, Publishes, push_now(Session)}
+    end;
```

Comment:
Nit:

```suggestion
    {Publishes, Session} = drain_buffer(Session1),
    {ok, Publishes, push_now(Session)};
```

Reply:
Thanks. Originally there was an "optimization" to avoid push whenever there is nothing to drain, but it broke the push loop in some cases.